Metadata-Matcher/run_script.py

```diff
@@ -0,0 +1,168 @@
+        """
+        Match metadata and text files
+        """
+        logging.info("Starting metadata matching")
+
+        # Enumerate all text files
+        for i, text_file in enumerate(self.files):
+            logging.info(
+                f"Scanning File ({i + 1}/{len(self.files)}) {text_file}")
+            match_found = False
+
+            # Enumerate each row in metadata sheet
+            for article_meta in self.sheet.articles:
+                cutoff = 70 if self.use_abstract else 50
+                score = self._get_score_for_text(text_file, article_meta)
+                if score >= cutoff:
+                    logging.info("Found match")
+                    self._found_match(text_file, article_meta, score)
+                    match_found = True
+                    break
+
+            if not match_found:
+                logging.info("No match found")
+                self._copy_file_not_found(text_file)
+
+        logging.info(
+            f"Saving matched files metadata to '{self.destination_found}'")
+        self.sheet.save(self.destination_found)
+        logging.info(
+            f"Successfully saved matched files metadata to '{self.destination_found}'")
+
+    def _get_score_for_text(self, text_file, article_meta):
+        # Match contents of `text_file` and `article_meta`
+        with open(text_file, encoding='utf8') as f:
+            size_to_read = self._get_file_size_to_read(text_file)
+            data = f.read(size_to_read)
+            data_normalized = self._normalize(data)
+
+            pmid = article_meta["pmid"]
+            pmcid = article_meta["pmcid"]
+            abstract = article_meta["abstract"]
+            title = article_meta["title"]
+
+            # Scoring
+            score = 0
+            if title and self._normalize(title) in data_normalized:
+                score += 30
+            if pmid and pmid in data:
+                score += 5
+            if pmcid and pmcid in data:
+                score += 5
+            if self.use_abstract and abstract and self._normalize(abstract) in data_normalized:
+                score += 40
+
+            if len(article_meta["authors"]) > 0:
+                # Match all authors
+                author_matches = 0
+                for author in article_meta["authors"]:
+                    first_name = self._normalize(author[0])
+                    last_name = self._normalize(author[1])
+
+                    if f"{last_name}{first_name}" in data_normalized:
+                        author_matches += 1
+                    elif f"{first_name}{last_name}" in data_normalized:
+                        author_matches += 1
+
+                author_match_percentage = author_matches / \
+                    len(article_meta["authors"])
+                score += 20 * author_match_percentage
+
+            return score
+
+    def _found_match(self, text_file, article_meta, score):
+        with open(text_file, encoding='utf8') as f:
```

Comment:
Although hopefully all files will be in utf-8 there might be a case where they are not. So I would make the encoding a configuration property so that if someone has a dataset with a different encoding they can just change it.

Reply:
might not be a bad idea to use chardet to try and figure out the encoding as well.